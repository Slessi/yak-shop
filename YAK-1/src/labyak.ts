export interface LabYak {
  name: string;
  age: number;
  sex: string;
}
