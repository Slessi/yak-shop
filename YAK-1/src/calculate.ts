import { LabYak } from "./labyak";

interface LabYakResult {
  name: string;
  age: number;
  milk: number;
  wool: number;
  lastShave: number;
}

export function calculate(labyaks: LabYak[], totalDays: number) {
  const results = new Map<number, LabYakResult>(
    labyaks.map((l, i) => [
      i,
      {
        name: l.name,
        age: l.age,
        milk: 0,
        wool: 0,
        lastShave: 0,
      },
    ])
  );

  let currentDay = 0;

  while (currentDay < totalDays) {
    results.forEach((result) => {
      // D = age in days
      const D = result.age * 100;

      // A LabYak dies the day he turns 10.
      // Using .toFixed to prevent floating point errors
      if (result.age.toFixed(2) === "10.00") {
        return;
      }

      // Each day a LabYak produces 50-D*0.03 liters of milk
      result.milk += 50 - D * 0.03;

      if (canShaveLabYak(result, currentDay)) {
        result.wool += 1;
        result.lastShave = currentDay;
      }

      result.age += 0.01;
    });

    currentDay++;
  }

  return results;
}

function canShaveLabYak({ age, lastShave }: LabYakResult, currentDay: number) {
  // A yak can be first shaven when he is 1 year.
  if (age < 1) {
    return false;
  }

  // The moment you open up the YakShop webshop will be day 0, and all yaks are eligible to be shaven
  if (currentDay === 0) {
    return true;
  }

  // At most every 8+age (D*0.01) days you can again shave a LabYak
  return currentDay - lastShave > 8 + age;
}
