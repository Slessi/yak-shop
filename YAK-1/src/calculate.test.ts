import { calculate } from "./calculate";
import { LabYak } from "./labyak";

describe("calculate", () => {
  const mockHerd: LabYak[] = [
    { name: "Betty-1", age: 4, sex: "f" },
    { name: "Betty-2", age: 8, sex: "f" },
    { name: "Betty-3", age: 9.5, sex: "f" },
  ];

  it.each<Parameters<typeof calculate>>([
    [mockHerd, 13],
    [mockHerd, 14],
    [mockHerd, 200],
  ])("calculates expected values", (...input) => {
    expect(calculate(...input)).toMatchSnapshot();
  });
});
