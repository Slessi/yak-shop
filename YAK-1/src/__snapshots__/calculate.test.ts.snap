// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculate calculates expected values 1`] = `
Map {
  0 => Object {
    "age": 4.129999999999997,
    "lastShave": 0,
    "milk": 491.66,
    "name": "Betty-1",
    "wool": 1,
  },
  1 => Object {
    "age": 8.129999999999997,
    "lastShave": 0,
    "milk": 335.65999999999997,
    "name": "Betty-2",
    "wool": 1,
  },
  2 => Object {
    "age": 9.629999999999997,
    "lastShave": 0,
    "milk": 277.16,
    "name": "Betty-3",
    "wool": 1,
  },
}
`;

exports[`calculate calculates expected values 2`] = `
Map {
  0 => Object {
    "age": 4.139999999999997,
    "lastShave": 13,
    "milk": 529.27,
    "name": "Betty-1",
    "wool": 2,
  },
  1 => Object {
    "age": 8.139999999999997,
    "lastShave": 0,
    "milk": 361.27,
    "name": "Betty-2",
    "wool": 1,
  },
  2 => Object {
    "age": 9.639999999999997,
    "lastShave": 0,
    "milk": 298.27000000000004,
    "name": "Betty-3",
    "wool": 1,
  },
}
`;

exports[`calculate calculates expected values 3`] = `
Map {
  0 => Object {
    "age": 5.999999999999957,
    "lastShave": 189,
    "milk": 7003.000000000014,
    "name": "Betty-1",
    "wool": 15,
  },
  1 => Object {
    "age": 9.999999999999957,
    "lastShave": 193,
    "milk": 4603.000000000013,
    "name": "Betty-2",
    "wool": 12,
  },
  2 => Object {
    "age": 9.99999999999999,
    "lastShave": 36,
    "milk": 1038.2500000000007,
    "name": "Betty-3",
    "wool": 3,
  },
}
`;
